src/uu/dd/src/dd.rs

```diff
@@ -0,0 +1,1296 @@
+        .get_appropriate_unit(false)
+        .format(1);
+
+    format!(
+        "{} bytes ({}, {}) copied, {:.1} s, {}/s",
+        update.bytes_total,
+        btotal_metric,
+        btotal_bin,
+        update.duration.as_secs_f64(),
+        xfer_rate
+    )
+}
+fn reprint_prog_line(update: &ProgUpdate) {
+    eprint!("\r{}", make_prog_line(update));
+}
+fn print_prog_line(update: &ProgUpdate) {
+    eprintln!("{}", make_prog_line(update));
+}
+fn print_xfer_stats(update: &ProgUpdate) {
+    print_io_lines(update);
+    print_prog_line(update);
+}
+
+/// Generate a progress updater that tracks progress, receives updates, and responds to signals.
+fn gen_prog_updater(rx: mpsc::Receiver<ProgUpdate>, xfer_stats: Option<StatusLevel>) -> impl Fn() {
+    // --------------------------------------------------------------
+    fn posixly_correct() -> bool {
+        env::var("POSIXLY_CORRECT").is_ok()
+    }
+    // --------------------------------------------------------------
+    move || {
+        const SIGUSR1_USIZE: usize = signal::SIGUSR1 as usize;
+
+        let sigval = Arc::new(AtomicUsize::new(0));
+
+        // TODO: SIGINFO seems to only exist for BSD (and therefore MACOS)
+        // I will probably want put this behind a feature-gate and may need to pass the value to handle as my own constant.
+        // This may involve some finagling with the signals library.
+        // see -> https://unix.stackexchange.com/questions/179481/siginfo-on-gnu-linux-arch-linux-missing
+        // if let Err(e) = signal_hook::flag::register_usize(signal::SIGINFO, sigval.clone(), signal::SIGINFO as usize)
+        // {
+        //     debug_println!("Internal dd Warning: Unable to register SIGINFO handler \n\t{}", e);
+        // }
+        if !posixly_correct() {
+            if let Err(e) =
+                signal_hook::flag::register_usize(signal::SIGUSR1, sigval.clone(), SIGUSR1_USIZE)
+            {
+                debug_println!(
+                    "Internal dd Warning: Unable to register SIGUSR1 handler \n\t{}",
+                    e
+                );
+            }
+        }
+
+        loop {
+            // Wait for update
+            let update = match (rx.recv(), xfer_stats) {
+                (Ok(update), Some(StatusLevel::Progress)) => {
+                    reprint_prog_line(&update);
+
+                    update
+                }
+                (Ok(update), _) => update,
+                (Err(_), _) =>
+                // recv only fails permanently
+                {
+                    break
+                }
+            };
+            // Handle signals
+            #[allow(clippy::single_match)]
+            match sigval.load(Ordering::Relaxed) {
+                SIGUSR1_USIZE => {
+                    print_xfer_stats(&update);
+                }
+                // SIGINFO_USIZE => ...
+                _ => { /* no signals recv'd */ }
+            };
```

Comment:
Instead of a match with one arm you could use `if let`:
```suggestion
            if let SIGNUSR1_USIZE = sigval.load(Ordering::Relaxed) {
                print_xfer_stats(&update);
            };
```
Although I guess you're keeping it as a `match` to allow for more signals later?

Reply:
That was the idea behind the match case yeah, but I think you bring up a good point. I removed the roughed out implementation for SIGINFO (the match case and a couple other lines). I think any future implementer will be just as likely to be confused by it than helped by it. A good suggestion.